services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/utils/SubnetManagementUtil.java

```diff
@@ -66,4 +76,134 @@ public static boolean checkSubnetRequestResourceIsValid(SubnetRequestWebJson res
+            return false;
+        }
+
+        // gatewayIP format is valid but it used the lowest base ip (xx.xx.xx.0) case and use the highest base ip (xx.xx.xx.255))
+        boolean isCidrValid = subnetService.verifyCidrBlock(cidr);
+        if (!isCidrValid) {
+            throw new FallbackException("cidr is invalid : " + cidr);
+        }
+
+        String[] ips = subnetService.cidrToFirstIpAndLastIp(cidr);
+        if (ips == null || ips.length != 2) {
+            throw new FallbackException("cidr transfer to first/last ip failed");
+        }
+
+        if (!checkIpIsInRange(gatewayIp, ips[0], ips[1])) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public static boolean checkGatewayIpIsInAllocatedRange(String gatewayIp, String cidr) {
+        if (gatewayIp == null) {
+            return false;
+        }
+        if (gatewayIp.length() == 0) {
+            return false;
+        }
+        String[] ips = subnetService.cidrToFirstIpAndLastIp(cidr);
+
+        long gatewayIpNum = getIpNum(gatewayIp);
+        long firstIpNum = getIpNum(ips[0]);
+        long lastIpNum = getIpNum(ips[1]);
+
+        if (gatewayIpNum < firstIpNum || gatewayIpNum > lastIpNum) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public static boolean checkIpIsValid(String gatewayIp) {
+        if (gatewayIp == null) {
+            return false;
+        }
+        String[] ips = gatewayIp.split("\\.");
+
+        if (ips == null || ips.length != 4) {
+            return false;
+        }
+
+        for (String segment : ips) {
+            char[] chars = segment.toCharArray();
+            for (int i = 0; i < chars.length; i ++) {
+                char c = chars[i];
+                if (c < '0' || c > '9') {
+                    return false;
+                }
+            }
+            Integer num = Integer.parseInt(segment);
+            if (num < 0 || num > 255) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean checkIpIsInRange(String gatewayIp, String firstIp, String lastIp) {
+        boolean isInnerIp = false;
+
+        long gatewayIpNum = getIpNum(gatewayIp);
+        long firstIpNum = getIpNum(firstIp);
+        long lastIpNum = getIpNum(lastIp);
+
+        isInnerIp = isInner(gatewayIpNum, firstIpNum - ConstantsConfig.LowIpInterval, lastIpNum + ConstantsConfig.HighIpInterval);
+
+        return isInnerIp;
+    }
+
+    private static long getIpNum(String ipAddress) {
+        String[] ip = ipAddress.split("\\.");
+        long a = Integer.parseInt(ip[0]);
+        long b = Integer.parseInt(ip[1]);
+        long c = Integer.parseInt(ip[2]);
+        long d = Integer.parseInt(ip[3]);
+        long ipNum = a * 256 * 256 * 256 + b * 256 * 256 + c * 256 + d;
+        return ipNum;
+    }
+
+    private static boolean isInner(long userIp, long begin, long end) {
+        return (userIp >= begin) && (userIp <= end);
```

Comment:
I think "userIp > begin" or the client of this method needs to pass a larger begin.

Reply:
If you've UT covered this, then ignore the above comments :-)

Reply:
okay ~